From 95ce01f3591ef10e7866629bf4d03976e0593d8e Mon Sep 17 00:00:00 2001
From: "chak-kei.lam" <chak-kei.lam@mediatek.com>
Date: Tue, 4 Nov 2025 14:02:01 +0800
Subject: [PATCH] net: ethernet: mtk_eth_soc: change qdma txq num to 32

Without this patch, the QDMA TXQ number is insufficient
when supporting multiple switch PPPQ.
---
 drivers/net/ethernet/mediatek/mtk_eth_soc.c | 9 ++++++---
 drivers/net/ethernet/mediatek/mtk_eth_soc.h | 2 +-
 2 files changed, 7 insertions(+), 4 deletions(-)

--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.c
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
@@ -991,7 +991,8 @@ static void mtk_set_queue_speed(struct m
 	}
 
 out:
-	ofs = MTK_QTX_OFFSET * idx;
+	mtk_w32(eth, (idx / MTK_QTX_PER_PAGE), soc->reg_map->qdma.page);
+	ofs = MTK_QTX_OFFSET * (idx % MTK_QTX_PER_PAGE);
 	old_val = mtk_r32(eth, soc->reg_map->qdma.qtx_sch + ofs);
 
 	if (mtk_is_netsys_v2_or_greater(eth))
@@ -3031,6 +3032,8 @@ static int mtk_tx_alloc(struct mtk_eth *
 		mtk_w32(eth, ring->last_free_ptr, soc->reg_map->qdma.drx_ptr);
 
 		for (i = 0, ofs = 0; i < MTK_QDMA_NUM_QUEUES; i++) {
+			mtk_w32(eth, (i / MTK_QTX_PER_PAGE), soc->reg_map->qdma.page);
+
 			val = (QDMA_RES_THRES << 8) | QDMA_RES_THRES;
 			mtk_w32(eth, val, soc->reg_map->qdma.qtx_cfg + ofs);
 
@@ -3054,7 +3057,7 @@ static int mtk_tx_alloc(struct mtk_eth *
 				val |= FIELD_PREP(MTK_QTX_SCH_TX_SEL_V2, 2);
 
 			mtk_w32(eth, val, soc->reg_map->qdma.qtx_sch + ofs);
-			ofs += MTK_QTX_OFFSET;
+			ofs = (ofs + MTK_QTX_OFFSET) % (MTK_QTX_OFFSET * MTK_QTX_PER_PAGE);
 		}
 		val = MTK_QDMA_TX_SCH_MAX_WFQ | (MTK_QDMA_TX_SCH_MAX_WFQ << 16);
 		mtk_w32(eth, val, soc->reg_map->qdma.tx_sch_rate);
--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.h
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.h
@@ -26,7 +26,7 @@
 #define MTK_MAX_DSA_PORTS	7
 #define MTK_DSA_PORT_MASK	GENMASK(2, 0)
 
-#define MTK_QDMA_NUM_QUEUES	16
+#define MTK_QDMA_NUM_QUEUES	32
 #define MTK_QDMA_PAGE_SIZE	2048
 #define MTK_MAX_RX_LENGTH	1536
 #define MTK_MAX_RX_LENGTH_2K	2048
--- a/drivers/net/ethernet/mediatek/mtk_eth_dbg.c
+++ b/drivers/net/ethernet/mediatek/mtk_eth_dbg.c
@@ -160,7 +160,7 @@ static ssize_t qdma_sched_read(struct fi
 	for (i = 0; i < MTK_QDMA_NUM_QUEUES; i++) {
 		mtk_w32(eth, (i / MTK_QTX_PER_PAGE), soc->reg_map->qdma.page);
 		sch_reg = mtk_r32(eth, soc->reg_map->qdma.qtx_sch +
-				       (id % MTK_QTX_PER_PAGE) * MTK_QTX_OFFSET);
+				       (i % MTK_QTX_PER_PAGE) * MTK_QTX_OFFSET);
 		if (mtk_is_netsys_v2_or_greater(eth))
 			scheduler = FIELD_GET(MTK_QTX_SCH_TX_SEL_V2, sch_reg);
 		else
